[MRG] Make random_state descriptions more informative and refer to Glossary for clusters module #10548

shivamriky opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
#10548 

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
Updated documentation for random_state in file 'sklearn/cluster/k_means_.py' to remove redundancy. Now it refers the glossary for more details with more specific information within the code. Partial fix for issue 10548.


#### Any other comments?
The issue #10548 should not be closed unless other scripts containing the random_state variable are updated. Therefore, I have avoided the use of '#\<issue number\>'

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slowto
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

shivaam committed:
Updated documentation for random_state to remove redundancy. Now it refers the glossary for more details with more specific information within the code. Fix for issue #10548.

jnothman commented on sklearn/cluster/k_means_.py:
```diff
@@ -222,11 +222,10 @@ def k_means(X, n_clusters, init='k-means++', precompute_distances='auto',
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Determines random number generation for centroid initialization. It
+        also generates random seeds for the parallel execution.
```
I'm not sure this means anything for the user, if those seeds are then used only for init

jnothman commented on sklearn/cluster/k_means_.py:
```diff
@@ -1046,11 +1039,10 @@ def _mini_batch_step(X, x_squared_norms, centers, counts,
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Determines random number generation for centroid initialization and to
+        pick new clusters amongst observations with uniform probability
```
Needs .

shivamriky replied:
Do you mean 'Needs' instead of 'Determines'?

jnothman replied:
It needs a `.`

shivamriky replied:
Definitely, it needs a '.' . I was too careless about the grammar will keep it in mind. 

Thank you.

jnothman reviewed (commented):
Please always give your pr a descriptive title that stands alone. Please update it. Also refer to issues with a #.

But this looks like a good start. Yes, try all of cluster

jnothman commented:
Please add another  commit to fix those things up.

shivaam committed:
Fixed issues raised in the merge request. Fix for issue #10548

shivamriky commented:
Please review.

jnothman commented:
To quell @amueller's concerns at #10548, we may need to add in the most important case: "Use an int for reproducibility."

jnothman reviewed (commented):
Otherwise this is good

shivamriky commented:
'Use an int for reproducibility' - do you mean that an integer value can be passed to reproduce same set of random values at a later stage ?

I couldn't find @amueller's comment. But do you want to add the line --

'Use an int for reproducibility.' in all the comments?

jnothman commented:
https://github.com/scikit-learn/scikit-learn/issues/10415#issuecomment-367106812 is what I meant

Your wording may be better. I'm just trying to make these descriptions as helpful as possible

shivaam committed:
Added description of using an int value with random_state to reproduce the output. #10548

shivamriky commented:
Done.

shivamriky commented:
Let me know if the comment looks fine. I will do the same for all other scripts..

jnothman commented on sklearn/cluster/k_means_.py:
```diff
@@ -222,11 +223,9 @@ def k_means(X, n_clusters, init='k-means++', precompute_distances='auto',
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Determines random number generation for centroid initialization.
```
Reproducibility missing here.

jnothman commented on sklearn/cluster/k_means_.py:
```diff
@@ -56,7 +56,8 @@ def _k_init(X, n_clusters, x_squared_norms, random_state, n_local_trials=None):
     random_state : RandomState
-        The generator used to initialize the centers.
+        The generator used to initialize the centers. Pass an int for
+        reproducing the random values.
```
I don't like how the wording has turned out. We could say "to reproduce", but "for reproducing" is awkward.

"Use an integer to make the randomness deterministic"??

shivamriky replied:
Sure, I will update to the following:
 **Use an int to make the randomness deterministic**

jnothman commented:
I think that's fine

shivaam committed:
Updated the description for using an int value with random_state to reproduce the output. #10548

shivaam committed:
1. Updated k_means.py modified one comment as random state
is not default for _k_init function.
2. Updated the documentation for random_state in mean_shift_.py #10548

shivamriky commented:
I think the jobs just timed out. Can anyone help me with the error? I am not able to find the problem with the code.

shivamriky commented:
Still can't figure out why the tests failed. Is the only way to re-run ?

shivaam committed:
Updated documentation for random_state for biclustering for issue #10548

shivaam committed:
Updated the documentation of random state in spectral clustering #10548

shivamriky commented:
I have updated all the documentation. Please review whenever you can and now we can close this PR if all good.

shivaam committed:
Fixing continouos integration error - a comment was too long #10548

jnothman commented on sklearn/cluster/bicluster.py:
```diff
@@ -238,11 +238,9 @@ class SpectralCoclustering(BaseSpectral):
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Use an int to make the randomness deterministic.
```
Do we want to describe what is randomised here?

jnothman commented on sklearn/cluster/bicluster.py:
```diff
@@ -370,11 +368,9 @@ class SpectralBiclustering(BaseSpectral):
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Use an int to make the randomness deterministic.
```
Do we want to describe what is randomised here?

shivamriky commented:
There is very little information on bicluster algorithm. Plus the comments in the file didn't help much. I will try to understand the algorithm during the weekend. Let me know if you have any ideas.

jnothman commented:
Seems it just randomizes the decomposition​

shivaam committed:
Updated the documentation for random_state in bicluster.py #10548

shivamriky commented:
Let me know once u have had a look :)

shivamriky commented:
@jnothman Just checking if you have other comments so that I can fix over the weekend?

jnothman reviewed (approved):
@shivamriky, sorry, I've been very busy lately and trying to focus on more important things. This seems good to me at a glance.

TomDLT commented on sklearn/cluster/bicluster.py:
```diff
@@ -238,11 +238,10 @@ class SpectralCoclustering(BaseSpectral):
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Used for randomizing the singular value decomposition. Use an int
```
Used for randomizing the singular value decomposition and the k-means initialization.

TomDLT reviewed (approved):
Here are some comments, otherwise LGTM

shivamriky commented on sklearn/cluster/spectral.py:
```diff
@@ -38,11 +38,9 @@ def discretize(vectors, copy=True, max_svd_restarts=30, n_iter_max=20,
-        If None, the random number generator is the RandomState instance used
-        by `np.random`.
+    random_state : int, RandomState instance or None (default)
+        Use an int to make the randomness deterministic.
```
@TomDLT @jnothman  thank you for approval. 

@TomDLT can I make the required changes and commit in the same branch?  Sorry, I am not very sure about this process.

TomDLT replied:
Yes, just commit to the same branch and push to your fork as before.

shivaam committed:
Incorporated additional review comments #10548

shivamriky commented:
@TomDLT Updated the code. I think it is ready to be merged now. Let me know in case anything else would be required.

TomDLT commented:
Merged !  Thanks @shivamriky !

shivamriky commented on sklearn/cluster/bicluster.py:
```diff
@@ -238,11 +238,10 @@ class SpectralCoclustering(BaseSpectral):
+    random_state : int, RandomState instance or None (default)
+        Used for randomizing the singular value decomposition and the k-means
+        initialization. Use an int to make the randomness deterministic.
+        See :term:`random_state <Glossary>`.
```
Thank you for updating that. Any particular reason to use it in that particular way?

qinhanmin2014 replied:
I can't give a particular reason but I think it's a grammatical rule. Generally I'll refer to the rest of the project after finding some problems in the doc. You might Google if you're interested in it.
